Upgrade to HQRM, New Resource, Features and Fixes - Fixes #53

PlagueHO opened the pull request:
This PR addresses:
- Fixes #58
- Fixes #57 
- Fixes #56 
- Fixes #54 
- Fixes #53 
- Fixes #51 
- Fixes #47 
- Fixes #35 
- Fixes #24 
- Fixes #22
- Fixes #20 

This PR contains the following changes:
* Updated to meet HQRM guidelines.
* Moved all strings into localization files.
* Fixed examples to import xStorage module.
* Fixed Readme.md layout issues.
* xWaitForDisk:
  - Added support for setting DriveLetter parameter with or without colon.
  - MOF Class version updated to 1.0.0.0.
* xWaitForVolume:
  - Added new resource.
* xStorageCommon:
  - Added helper function module.
* xDisk:
  - Added validation of DriveLetter parameter.
  - Added support for setting DriveLetter parameter with or without colon.
  - Removed obfuscation of drive/partition errors by eliminating try/catch block.
  - Improved code commenting.
  - Reordered tests so they are in same order as module functions to ease creation.
  - Added FSFormat parameter to allow disk format to be specified.
  - Size or AllocationUnitSize mismatches no longer trigger Set-TargetResource because these values can't be changed (yet).
  - MOF Class version updated to 1.0.0.0.
* xMountImage (Breaking Change):
  - Removed Name parameter (Breaking Change)
  - Added validation of DriveLetter parameter.
  - Added support for setting DriveLetter parameter with or without colon.
  - MOF Class version updated to 1.0.0.0.
  - Enabled mounting of VHD/VHDx/VHDSet disk images.
  - Added StorageType and Access parameters to allow mounting VHD and VHDx disks as read/write.
* xDiskAccessPath:
  - Added new resource.

As usual, I apologize in advance to any reviewers for the size of this PR :grin: and also any mistakes I have made (no doubt there are a few).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xstorage/60)
<!-- Reviewable:end -->

msftclas commented:
Hi __@PlagueHO__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

narrieta commented:
Reviewed 4 of 45 files at r1.
Review status: 4 of 45 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[README.md, line 16 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmT77RQZu2tIIq0gS_:-KSmT77RQZu2tIIq0gSa:bfcnc0i) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/README.md#L16)):*
>  the latest version of PowerShell (v4.0, 

The latest version is now 5.0

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 38 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmUuZ5ybdkbMsWMZL3:-KSmUuZ5ybdkbMsWMZL4:bs73b8h) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L38)):*
> if required

Do you mean 'if required' or 'if provided'?

Same in Set

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 78 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmVOqUJ8IfqQDkewhw:-KSmVOqUJ8IfqQDkewhx:bwm1iae) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L78)):*
>  -ErrorAction SilentlyContinue

if we continue on error $disk may be $null and then we'll have problems when the return value tries to access $disk.Number. Also, since we are silently continuing there will be no indication of where the problem is coming from.

NOTE: I just realized this is coming from the previous version of the code, so if this is too time consuming to fix now we should open an issue. There are more instances below.

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 213 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmYvfDR5CPRlZ_RO5-:-KSmYvfDR5CPRlZ_RO50:bhsjsdo) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L213)):*
> ```PowerShell
>             $disk | Initialize-Disk `
>                 -PartitionStyle "GPT"
>         } # "RAW"
> ```

add 'break' at the end of each script block so that control flow does not fall into next case.

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 340 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmaHZeJqq0cWrRrlbw:-KSmaHZeJqq0cWrRrlbx:b-rmw3xr) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L340)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>
>                 Write-Verbose -Message ( @(
>                         "$($MyInvocation.MyCommand): "
>                         $($LocalizedData.FileSystemFormatMismatch -f `
>                             $DriveLetter,$fileSystem,$FSFormat)
>                     ) -join '' )
</details>

This is more of a wish-list item for future coding: consider writing a utility function to prepend the invocation info to the message. 

Not only this would make the code easier to read, but also would make implementing changes to the format of the message a lot easier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Reviewed 2 of 45 files at r1.
Review status: 6 of 45 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1, line 4 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmbakcTQlCXv6wfwr8:-KSmbakcTQlCXv6wfwr9:bd7xgjp) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1#L4)):*
>  is
typo(?): remove "is"

---

*[DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1, line 21 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmbnaC5WiO3Foy6J5w:-KSmbnaC5WiO3Foy6J5x:bfy9mec) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1#L21)):*
> initialised 
initialized

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Review status: 2 of 45 files reviewed at latest revision, 7 unresolved discussions.

---

*[README.md, line 16 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmT77RQZu2tIIq0gS_:-KSp1jTkptK77fjjkOZs:b-839tjs) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/README.md#L16)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> >  the latest version of PowerShell (v4.0, 
> 
> The latest version is now 5.0
</details>
Oh too true.

I'd actually argue that this whole section doesn't need to be included at all (it isn't in other resources). DSC isn't supported on anything under WMF4.0 anyway. Would you object to the requirements section being removed?

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 38 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmUuZ5ybdkbMsWMZL3:-KSp3FSmcIdxvcQMl7wG:b1mwrme) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L38)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > if required
> 
> Do you mean 'if required' or 'if provided'?
> 
> Same in Set
</details>
I did mean 'if required'. But agree this isn't clear. I have changed this to clarify it.

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 78 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmVOqUJ8IfqQDkewhw:-KSp4BmNwLU8RMTPonLx:b-6sjiz) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L78)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> >  -ErrorAction SilentlyContinue
> 
> if we continue on error $disk may be $null and then we'll have problems when the return value tries to access $disk.Number. Also, since we are silently continuing there will be no indication of where the problem is coming from.
> 
> NOTE: I just realized this is coming from the previous version of the code, so if this is too time consuming to fix now we should open an issue. There are more instances below.
</details>
I'll take a look and see what will be required to fix this now. I'll spend a bit of time on this over the weekend.

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 213 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmYvfDR5CPRlZ_RO5-:-KSp4ppfQ6HlgLAn4RgJ:b-2ntk0l) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L213)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> add 'break' at the end of each script block so that control flow does not fall into next case.
</details>
Done. Good catch. Also done in xDiskAccessPath

---

*[DSCResources/MSFT_xDisk/MSFT_xDisk.psm1, line 340 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmaHZeJqq0cWrRrlbw:-KSp4y9fW_5ZeezMI2yr:b-8gm1tu) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/MSFT_xDisk.psm1#L340)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> >                 Write-Verbose -Message ( @(
> >                         "$($MyInvocation.MyCommand): "
> >                         $($LocalizedData.FileSystemFormatMismatch -f `
> >                             $DriveLetter,$fileSystem,$FSFormat)
> >                     ) -join '' )
> 
> This is more of a wish-list item for future coding: consider writing a utility function to prepend the invocation info to the message. 
> 
> Not only this would make the code easier to read, but also would make implementing changes to the format of the message a lot easier.
</details>
Very good point! I'll keep this in mind and ensure I include it in future updates.

---

*[DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1, line 4 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmbakcTQlCXv6wfwr8:-KSp5TvElEbV_lL5j96b:b-896fix) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1#L4)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> >  is
> typo(?): remove "is"
</details>
Done.

---

*[DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1, line 21 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmbnaC5WiO3Foy6J5w:-KSp5_YYQOib0YM4opqc:b-n4ul7a) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xDisk/en-us/MSFT_xDisk.strings.psd1#L21)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > initialised 
> initialized
</details>
Sorry, slipped back into en-nz . Good catch. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Reviewed 3 of 45 files at r1.
Review status: 5 of 45 files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 16 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmT77RQZu2tIIq0gS_:-KSqhl6QxfYsDEXnknQf:b-5tltab) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/README.md#L16)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary>
> Oh too true.
> 
> I'd actually argue that this whole section doesn't need to be included at all (it isn't in other resources). DSC isn't supported on anything under WMF4.0 anyway. Would you object to the requirements section being removed?
</details>
Yup, we should remove it. Thanks!

---

*[DSCResources/MSFT_xMountImage/MSFT_xMountImage.psm1, line 54 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSn-ns5Xu7rwLeZxBJ-:-KSn-ns6qgYvRKpF9h2r:byt4z3w) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xMountImage/MSFT_xMountImage.psm1#L54)):*
> [Microsoft.PowerShell.Cmdletization.GeneratedTypes.DiskImage.StorageType]

Why is this needed? Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Reviewed 7 of 45 files at r1, 3 of 5 files at r2.
Review status: 15 of 45 files reviewed at latest revision, 6 unresolved discussions.

---

*[DSCResources/MSFT_xWaitForDisk/en-us/MSFT_xWaitForDisk.strings.psd1, line 10 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqm71WCcRXClQdntVK:-KSqm71WCcRXClQdntVL:b-pm7fvo) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xWaitForDisk/en-us/MSFT_xWaitForDisk.strings.psd1#L10)):*
> Tetting 
Testing

---

*[DSCResources/StorageCommon/StorageCommon.psm1, line 14 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqmOCl8xLCf-cyx6N0:-KSqmOCl8xLCf-cyx6N1:bb3agw1) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/StorageCommon/StorageCommon.psm1#L14)):*
> #region localizeddata
> if (Test-Path "${PSScriptRoot}\${PSUICulture}")
> {
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>     Import-LocalizedData `
>         -BindingVariable LocalizedData `
>         -Filename StorageCommon.strings.psd1 `
>         -BaseDirectory "${PSScriptRoot}\${PSUICulture}"
</details>
> }
> else
> {
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>
>     #fallback to en-US
>     Import-LocalizedData `
>         -BindingVariable LocalizedData `
>         -Filename StorageCommon.strings.psd1 `
>         -BaseDirectory "${PSScriptRoot}\en-US"
</details>
> }
> #endregion

Adding to the wish list :) - We should encapsulate this in a common function that can be used by all tests that have localized data

---

*[DSCResources/StorageCommon/StorageCommon.psm1, line 97 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqo4zh29wFhjGbhoA1:-KSqo4zh29wFhjGbhoA2:bd2q86f) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/StorageCommon/StorageCommon.psm1#L97)):*
> Test

Minor: The Test verb usually indicates a boolean return value (instead of throwing or returning a normalized value).

Since this is an internal module (not user visible), this is not the biggest priority

---

*[DSCResources/StorageCommon/StorageCommon.psm1, line 112 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqmvPb6duUy7bVQDkO:-KSqmvPb6duUy7bVQDkP:b-jj3n4m) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/StorageCommon/StorageCommon.psm1#L112)):*
> ^([A-Za-z]{1})[:]?$

Minor: {1} and [] are redundant:  ^([A-Za-z]):?$

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Reviewed 9 of 45 files at r1.
Review status: 23 of 45 files reviewed at latest revision, 10 unresolved discussions.

---

*[Tests/Integration/MSFT_xDisk.Integration.Tests.ps1, line 44 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqtlFHvddCcQJDKGHL:-KSqtlFI0APFVDuYRQOd:b-7d1uj5) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDisk.Integration.Tests.ps1#L44)):*
> # Unsupported OS type for testing
We should output this a verbose message

(same for other integration tests)

---

*[Tests/Integration/MSFT_xDisk.Integration.Tests.ps1, line 52 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqtptaixmgLj6xYCr-:-KSqtptaixmgLj6xYCr0:bxhm3au) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDisk.Integration.Tests.ps1#L52)):*
>  Break

return?

(same for other integration tests)

---

*[Tests/Integration/MSFT_xDisk.Integration.Tests.ps1, line 60 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqw1wiXCFqTJBw_X0w:-KSqw1wiXCFqTJBw_X0x:b-g4hnir) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDisk.Integration.Tests.ps1#L60)):*
> BeforeAll

Minor - Instead of setting the scenario in BeforeAll, I'd use Context with a brief description of the scenario: Context 'When adding a drive letter to a disk'.

This gives the reader a better idea of what the test is doing and the output of Pester will also give similar info.

(same for other integration tests)

---

*[Tests/Integration/MSFT_xDiskAccessPath.Integration.Tests.ps1, line 23 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqxxtSvWnlbXfpt4TD:-KSqxxtSvWnlbXfpt4TE:b-xm1w3c) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDiskAccessPath.Integration.Tests.ps1#L23)):*
> ```PowerShell
> try
> {
>     # Ensure that the tests can be performed on this computer
> ```

There seems to be a lot of commonality between some integration tests. Consider refactoring it into a common module.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Reviewed 8 of 45 files at r1.
Review status: 31 of 45 files reviewed at latest revision, 14 unresolved discussions.

---

*[Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1, line 1 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr0HmgHpDi1RbbXHND:-KSr0HmgHpDi1RbbXHNE:b-1w1n09) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1#L1)):*
> an basic
a basic

---

*[Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1, line 2 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr0Cnf3fUkaVxZb8JQ:-KSr0Cnf3fUkaVxZb8JR:by23b6l) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1#L2)):*
> fil emust

file must

---

*[Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1, line 4 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr0S1qGFtdAYmxv6PZ:-KSr0S1qGFtdAYmxv6P_:bn7095q) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1#L4)):*
> ```PowerShell
> # in the same folder as this file. The ISO fil emust be a valid ISO file that can
> # normally be mounted. If the test.iso file is not found the tests will not run.
> ```

We should clarify this comment. When I read it I am not sure if this is something I need to do before running the test (no) or if the test sets this up (yes).

---

*[Tests/Integration/MSFT_xWaitForVolume.config.ps1, line 1 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr4NiiQVFrq2Q7GUsz:-KSr4NiiQVFrq2Q7GUt-:b-y6vbqq) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xWaitForVolume.config.ps1#L1)):*
> [PSObject] 

Minor: no need to use PSObject, the hashtable would work as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
Reviewed 7 of 45 files at r1.
Review status: 32 of 45 files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

narrieta commented:
I reviewed all files but the new resources, which should be reviewed by @kwirkykat, @TravisEz13, @KarolKaczmarek, @mbreakey3, @HemantMahawar, @grayzu

KarolKaczmarek commented:
Review status: 32 of 45 files reviewed at latest revision, 18 unresolved discussions.

---

*[DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1, line 85 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrnLnqti7RWxPuttJD:-KSrnLnqti7RWxPuttJE:b-efqjba) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1#L85)):*
> ```PowerShell
> 
>     # Prepare the AccessPath used in the CIM/WMI query (replaces '\' with '\\')
>     $queryAccessPath = $AccessPath -replace '\\','\\'
> ```

does it actually do anything?

---

*[DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1, line 240 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrk2I59r9I1UuCVDrm:-KSrk2I59r9I1UuCVDrn:bqwzhs7) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1#L240)):*
> ```PowerShell
>                     "$($MyInvocation.MyCommand): "
>                     $($LocalizedData.CreatingPartitionMessage `
>                         -f $DiskNumber,"$($Size/1kb) kb")
> ```

Technically we should display KB since those will be kilobytes, not kilobits. 

---

*[DSCResources/MSFT_xDiskAccessPath/en-us/MSFT_xDiskAccessPath.strings.psd1, line 23 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrp2gnLHDFFAPE0RM9:-KSrp2gnLHDFFAPE0RMA:b-lr7b8w) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xDiskAccessPath/en-us/MSFT_xDiskAccessPath.strings.psd1#L23)):*
> ```PowerShell
>     AccessPathNotFoundMessage = A volume assigned to access path '{0}' was not found.
>     SizeMismatchMessage = Volume assigned to access path '{0}' has size {1}, which does not match expected size {2}.
>     AllocationUnitSizeMismatchMessage = Volume assigned to access path '{0}' has allocation unit size {1} kb does not match expected allocation unit size {2} kb.
> ```

KB here as well

---

*[Examples/Sample_InitializeDataDiskWithAccessPath.ps1, line 11 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrt9bJLxxWQFscVVLo:-KSrt9bJLxxWQFscVVLp:b-noq5r0) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Examples/Sample_InitializeDataDiskWithAccessPath.ps1#L11)):*
> ```PowerShell
>     Node localhost
>     {
>         xWaitforDisk Disk2
> ```

Can we add sample for xWaitForVolume as well? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
@KarolKaczmarek, @narrieta - I think I've addressed everything. But I also made some additional changes and fixes that were picked up in xDisk when I added some additional unit tests.

---

Review status: 14 of 45 files reviewed at latest revision, 18 unresolved discussions.

---

*[README.md, line 16 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSmT77RQZu2tIIq0gS_:-KSxN67ZAQFX4uvmM0ZE:b-896fix) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/README.md#L16)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> Yup, we should remove it. Thanks!
</details>
Done.

---

*[DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1, line 85 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrnLnqti7RWxPuttJD:-KSxNDcdzZ35Rlcnscub:b-k8m279) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1#L85)):*
<details><summary><i>Previously, KarolKaczmarek (Karol Kaczmarek) wrote…</i></summary>
> does it actually do anything?
</details>
Yes - it does. :smile: The first parameter is a RegEx so the backslash is actually escaped. The second parameter is not a RegEx so does not need to be escaped. So the comment on the previous line is correct (replaces '\' with '\\')

---

*[DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1, line 240 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrk2I59r9I1UuCVDrm:-KSxOJdoMSHGpc8M3KKK:b-c4o3av) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xDiskAccessPath/MSFT_xDiskAccessPath.psm1#L240)):*
<details><summary><i>Previously, KarolKaczmarek (Karol Kaczmarek) wrote…</i></summary>
> Technically we should display KB since those will be kilobytes, not kilobits. 
</details>
Fixed (and all the other occurrences).

---

*[DSCResources/MSFT_xDiskAccessPath/en-us/MSFT_xDiskAccessPath.strings.psd1, line 23 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrp2gnLHDFFAPE0RM9:-KSxOP7fFzu5cy8ZMpxF:b-896fix) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xDiskAccessPath/en-us/MSFT_xDiskAccessPath.strings.psd1#L23)):*
<details><summary><i>Previously, KarolKaczmarek (Karol Kaczmarek) wrote…</i></summary>
> KB here as well
</details>
Done.

---

*[DSCResources/MSFT_xMountImage/MSFT_xMountImage.psm1, line 54 at r1](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSn-ns5Xu7rwLeZxBJ-:-KSxOgSbyroIexph7Rox:b-xz177l) ([raw file](https://github.com/powershell/xstorage/blob/a9bb992f5b6572c36f726a29b00012c1a8fd0416/DSCResources/MSFT_xMountImage/MSFT_xMountImage.psm1#L54)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > [Microsoft.PowerShell.Cmdletization.GeneratedTypes.DiskImage.StorageType]
> 
> Why is this needed? Thanks!
</details>
It is actually because $diskImage.StorageType contains an integer but casting it to an enum  of type [Microsoft.PowerShell.Cmdletization.GeneratedTypes.DiskImage.StorageType] will cause the enum string to be returned (which is what the Get-TargetResource must return).

---

*[DSCResources/MSFT_xWaitForDisk/en-us/MSFT_xWaitForDisk.strings.psd1, line 10 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqm71WCcRXClQdntVK:-KSxOrFlAQkFCyvklsDd:btt8hbd) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/MSFT_xWaitForDisk/en-us/MSFT_xWaitForDisk.strings.psd1#L10)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > Tetting 
> Testing
</details>
Fixed + all other occurrences of this misspelling.

---

*[DSCResources/StorageCommon/StorageCommon.psm1, line 14 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqmOCl8xLCf-cyx6N0:-KSxP-TlzcDAHFuLdRy_:b8qf9vh) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/StorageCommon/StorageCommon.psm1#L14)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > #region localizeddata
> > if (Test-Path "${PSScriptRoot}\${PSUICulture}")
> > {
> >     Import-LocalizedData `
> >         -BindingVariable LocalizedData `
> >         -Filename StorageCommon.strings.psd1 `
> >         -BaseDirectory "${PSScriptRoot}\${PSUICulture}"
> > }
> > else
> > {
> >     #fallback to en-US
> >     Import-LocalizedData `
> >         -BindingVariable LocalizedData `
> >         -Filename StorageCommon.strings.psd1 `
> >         -BaseDirectory "${PSScriptRoot}\en-US"
> > }
> > #endregion
> 
> Adding to the wish list :) - We should encapsulate this in a common function that can be used by all tests that have localized data
</details>
Good point. I'll add this to the list of things I'll implement in the next round :smile:

---

*[DSCResources/StorageCommon/StorageCommon.psm1, line 97 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqo4zh29wFhjGbhoA1:-KSxTkaHH72hVWpbV36o:b43dv2d) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/StorageCommon/StorageCommon.psm1#L97)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > Test
> 
> Minor: The Test verb usually indicates a boolean return value (instead of throwing or returning a normalized value).
> 
> Since this is an internal module (not user visible), this is not the biggest priority
</details>
Good point. I'll raise this as a separate issue to be repaired. Same goes for Test-AssignPath I expect

---

*[DSCResources/StorageCommon/StorageCommon.psm1, line 112 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqmvPb6duUy7bVQDkO:-KSxPWf_71h6Wlb2YgRd:b-cf40p8) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/DSCResources/StorageCommon/StorageCommon.psm1#L112)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > ^([A-Za-z]{1})[:]?$
> 
> Minor: {1} and [] are redundant:  ^([A-Za-z]):?$
</details>
Nice catch! Fixed.

---

*[Examples/Sample_InitializeDataDiskWithAccessPath.ps1, line 11 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSrt9bJLxxWQFscVVLo:-KSxQ9Sqqz1cBYMKUrcc:bnl4a27) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Examples/Sample_InitializeDataDiskWithAccessPath.ps1#L11)):*
<details><summary><i>Previously, KarolKaczmarek (Karol Kaczmarek) wrote…</i></summary>
> Can we add sample for xWaitForVolume as well? 
</details>
There was one already in Sample_xMountImage_MountVHD.ps1. But I've added another to Sample_xMountImage_MountISO.ps1 (I've corrected the name of this file as well).

I didn't think it made much sense as an example in it's own right. But I could add specific xWaitForVolume examples if you think it'll help.

---

*[Tests/Integration/MSFT_xDisk.Integration.Tests.ps1, line 44 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqtlFHvddCcQJDKGHL:-KSxR9I-HokO3xpWiuiW:b3jql4b) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDisk.Integration.Tests.ps1#L44)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > # Unsupported OS type for testing
> We should output this a verbose message
> 
> (same for other integration tests)
</details>
Good point - done.

---

*[Tests/Integration/MSFT_xDisk.Integration.Tests.ps1, line 52 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqtptaixmgLj6xYCr-:-KSxV37Sxpl47jIkchhp:br8uxfr) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDisk.Integration.Tests.ps1#L52)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> >  Break
> 
> return?
> 
> (same for other integration tests)
</details>
In this the break will do the same thing as the return, but I've changed it anyway because a return is much clearer.

---

*[Tests/Integration/MSFT_xDisk.Integration.Tests.ps1, line 60 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqw1wiXCFqTJBw_X0w:-KSx_nxmuIee_KSRF_9g:b-9fmn3c) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDisk.Integration.Tests.ps1#L60)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > BeforeAll
> 
> Minor - Instead of setting the scenario in BeforeAll, I'd use Context with a brief description of the scenario: Context 'When adding a drive letter to a disk'.
> 
> This gives the reader a better idea of what the test is doing and the output of Pester will also give similar info.
> 
> (same for other integration tests)
</details>
Good suggestion. Done.

---

*[Tests/Integration/MSFT_xDiskAccessPath.Integration.Tests.ps1, line 23 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSqxxtSvWnlbXfpt4TD:-KSxhnO_ufaBAF6m1Bjj:b-l4fpfr) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xDiskAccessPath.Integration.Tests.ps1#L23)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> There seems to be a lot of commonality between some integration tests. Consider refactoring it into a common module.
</details>
Ok. Done.

---

*[Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1, line 1 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr0HmgHpDi1RbbXHND:-KSxhr7B7n5d1tB2Re-U:b-896fix) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1#L1)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > an basic
> a basic
</details>
Done.

---

*[Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1, line 2 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr0Cnf3fUkaVxZb8JQ:-KSxa881NBWHjpqtgFwV:b-896fix) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1#L2)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > fil emust
> 
> file must
</details>
Done.

---

*[Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1, line 4 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr0S1qGFtdAYmxv6PZ:-KSxatvrTil9iR-q3rsk:bflr19t) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xMountImage_ISO.Integration.Tests.ps1#L4)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> We should clarify this comment. When I read it I am not sure if this is something I need to do before running the test (no) or if the test sets this up (yes).
</details>
I've tried to clarify this.

---

*[Tests/Integration/MSFT_xWaitForVolume.config.ps1, line 1 at r2](https://reviewable.io:443/reviews/powershell/xstorage/60#-KSr4NiiQVFrq2Q7GUsz:-KSxbUXVYH_c5cKk6PHd:bl8lja6) ([raw file](https://github.com/powershell/xstorage/blob/cd93ec1075178097e348500232024f547c98474e/Tests/Integration/MSFT_xWaitForVolume.config.ps1#L1)):*
<details><summary><i>Previously, narrieta (Norberto Arrieta) wrote…</i></summary>
> > [PSObject] 
> 
> Minor: no need to use PSObject, the hashtable would work as well
</details>
Good point. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60#-:-KSxhtLhiNhf5Sw88Con:b-kmm0sa)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
@kwirkykat , @narrieta , @KarolKaczmarek - are you guys waiting on anything from me with this one? 😁 

I'm looking at moving on to working through the issues/HQRM on some of the other DSC modules so just trying to get a few of these PR's closed off before I get started. Cheers!

narrieta commented:
@plagueho : Nothing on my side; I've ack'ed all your replies. Thanks for the update.

---

Review status: 14 of 46 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60#-:-KUNbgwcjlMB55TIbUpH:b-9jo2fc)*
<!-- Sent from Reviewable.io -->

KarolKaczmarek commented:
Reviewed 15 of 48 files at r1, 1 of 5 files at r2, 33 of 33 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

KarolKaczmarek commented:
@PlagueHO @kwirkykat Looks good on my side either 👍

kwirkykat commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[README.md, line 340 at r3](https://reviewable.io:443/reviews/powershell/xstorage/60#-KUxzIS5wE5UTXc7qeYV:-KUxzIS5wE5UTXc7qeYW:b-42923b) ([raw file](https://github.com/powershell/xstorage/blob/52e86672296cb9b90ccb267b26fe007585da8bef/README.md#L340)):*
> ```Markdown
> ```
> 
> ## Contributing
> ```

Duplicate Contributing section

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
Thank you for reviewing! Thanks @KwirkyKat - fixed the Contributing problem.

---

Review status: 45 of 46 files reviewed at latest revision, 1 unresolved discussion.

---

*[README.md, line 340 at r3](https://reviewable.io:443/reviews/powershell/xstorage/60#-KUxzIS5wE5UTXc7qeYV:-KV3pR_d9ICreYFLQO3g:b-896fix) ([raw file](https://github.com/powershell/xstorage/blob/52e86672296cb9b90ccb267b26fe007585da8bef/README.md#L340)):*
<details><summary><i>Previously, kwirkykat (Katie Keim) wrote…</i></summary>
> Duplicate Contributing section
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60#-:-KV3pE67ORGH75f7--92:b-1bcs19)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xstorage/60)*
<!-- Sent from Reviewable.io -->
